### ISSUE TITLE
Codec.Decode panic with client-go v0.17.0

### ISSUE THREAD

**Issue by howardjohn** [kubernetes/kubernetes#86394]
**What happened**:

I have a test case:
```go

func TestSimple(t *testing.T) {
	var fakeKubeconfigData = `apiVersion: v1
kind: Config
clusters:
- cluster:
    certificate-authority-data: UEhPTlkK
    server: https://192.168.1.1
  name: prod0
contexts:
- context:
    cluster: prod0
    user: prod0
  name: prod0
current-context: prod0
users:
- name: prod0
  user:
    auth-provider:
      name: gcp`

	out, x, err := latest.Codec.Decode([]byte(fakeKubeconfigData), nil, nil)
	fmt.Println(out, x, err)
}
```

This behaves different with different client-go versions.

With `kubernetes-1.16.0`, the test passes. This is using `k8s.io/client-go v0.0.0-20190918160344-1fbdaa4c8d90` in go.mod with `k8s.io/apimachinery v0.0.0-20190913080033-27d36303b655` in go.sum

With `v0.17.0`, this test panics. This is using `k8s.io/client-go v0.17.0` in go.mod with `k8s.io/apimachinery v0.17.0` in go.sum.

Panic logs:
```
--- FAIL: TestSimple (0.00s)
panic: assignment to entry in nil map [recovered]
        panic: assignment to entry in nil map

goroutine 21 [running]:
testing.tRunner.func1(0xc000162100)
        /usr/local/go/src/testing/testing.go:874 +0x3a3
panic(0x993700, 0xae33f0)
        /usr/local/go/src/runtime/panic.go:679 +0x1b2
k8s.io/client-go/tools/clientcmd/api/v1.Convert_Slice_v1_NamedCluster_To_Map_string_To_Pointer_api_Cluster(0xc0000c0720, 0xc0000ad410, 0xafd6a0, 0xc0000cadc0, 0x0, 0x0)
        /usr/local/google/home/howardjohn/go/pkg/mod/k8s.io/client-go@v0.17.0/tools/clientcmd/api/v1/conversion.go:35 +0x1cb
k8s.io/client-go/tools/clientcmd/api/v1.autoConvert_v1_Config_To_api_Config(0xc0000c06e0, 0xc0000ad3e0, 0xafd6a0, 0xc0000cadc0, 0x9203b0, 0xc0000f9600)
        /usr/local/google/home/howardjohn/go/pkg/mod/k8s.io/client-go@v0.17.0/tools/clientcmd/api/v1/zz_generated.conversion.go:273 +0xa0
k8s.io/client-go/tools/clientcmd/api/v1.Convert_v1_Config_To_api_Config(...)
        /usr/local/google/home/howardjohn/go/pkg/mod/k8s.io/client-go@v0.17.0/tools/clientcmd/api/v1/zz_generated.conversion.go:291
k8s.io/client-go/tools/clientcmd/api/v1.RegisterConversions.func7(0x9c7f00, 0xc0000c06e0, 0x9c7e40, 0xc0000ad3e0, 0xafd6a0, 0xc0000cadc0, 0xc00013d948, 0x4b2b49)
        /usr/local/google/home/howardjohn/go/pkg/mod/k8s.io/client-go@v0.17.0/tools/clientcmd/api/v1/zz_generated.conversion.go:69 +0x6f
k8s.io/apimachinery/pkg/conversion.(*Converter).doConversion(0xc0000acc60, 0x9c7f00, 0xc0000c06e0, 0x9c7e40, 0xc0000ad3e0, 0x0, 0xc00018cec0, 0xc00013da20, 0x7600000000000020, 0x9b98a0)
        /usr/local/google/home/howardjohn/go/pkg/mod/k8s.io/apimachinery@v0.17.0/pkg/conversion/converter.go:498 +0x6d4
k8s.io/apimachinery/pkg/conversion.(*Converter).Convert(0xc0000acc60, 0x9c7f00, 0xc0000c06e0, 0x9c7e40, 0xc0000ad3e0, 0x0, 0xc00018cec0, 0xaf2ae0, 0xc0000ad3e0)
        /usr/local/google/home/howardjohn/go/pkg/mod/k8s.io/apimachinery@v0.17.0/pkg/conversion/converter.go:470 +0x9d
k8s.io/apimachinery/pkg/runtime.(*Scheme).convertToVersion(0xc000142d90, 0x984500, 0xaf2b60, 0xc0000c06e0, 0xaf39a0, 0xef3540, 0x9c7f00, 0xc0000c06e0, 0xb08180, 0x9c7f00)
        /usr/local/google/home/howardjohn/go/pkg/mod/k8s.io/apimachinery@v0.17.0/pkg/runtime/scheme.go:566 +0x9d9
k8s.io/apimachinery/pkg/runtime.(*Scheme).UnsafeConvertToVersion(...)
        /usr/local/google/home/howardjohn/go/pkg/mod/k8s.io/apimachinery@v0.17.0/pkg/runtime/scheme.go:493
k8s.io/apimachinery/pkg/runtime.unsafeObjectConvertor.ConvertToVersion(0xc000142d90, 0xaf2b60, 0xc0000c06e0, 0xaf39a0, 0xef3540, 0x0, 0x0, 0xaf2b60, 0xc0000c06e0)
        /usr/local/google/home/howardjohn/go/pkg/mod/k8s.io/apimachinery@v0.17.0/pkg/runtime/helper.go:39 +0x5f
k8s.io/apimachinery/pkg/runtime/serializer/versioning.(*codec).Decode(0xc000144c80, 0xc000166000, 0x124, 0x140, 0x0, 0x0, 0x0, 0xc0000a9488, 0x44f628, 0x3b3a1b477ac37, ...)
        /usr/local/google/home/howardjohn/go/pkg/mod/k8s.io/apimachinery@v0.17.0/pkg/runtime/serializer/versioning/versioning.go:171 +0x34c
istio.io/temp-test.TestSimple(0xc000162100)
        /usr/local/google/home/howardjohn/go/src/istio.io/temp-test/example_test.go:30 +0x98
testing.tRunner(0xc000162100, 0xa5a278)
        /usr/local/go/src/testing/testing.go:909 +0xc9
created by testing.(*T).Run
        /usr/local/go/src/testing/testing.go:960 +0x350
FAIL    istio.io/temp-test      0.058s
FAIL
```

Basically it its creating an object to decode into, and this object has some maps. These are set to nil, so when we try to populate them we get a panic.

**What you expected to happen**: 

Behavior remains the same as previous release, or at least have some alternative that works here?

**How to reproduce it (as minimally and precisely as possible)**:

I pasted the test above that reproduces this. This can be reproduced with a project with just that file, and a go.mod file with only client-go. Using go 1.13
